### PR TITLE
ezLdap extension Ldap shadetree by dn

### PR DESCRIPTION
**Pull request by sfi2022**
**Description**
Zammad needs the possibility to synchronize ldap ou hierarchie (= lhm ldap shade tree in ApacheDirectoryStudio) and ou contained user to Zammad.

**Short description or comments**
The merge request contains a json convert including dependency. This can also be solved differently and is not necessary. We can take that out again.

**Reference**
 https://github.com/it-at-m/zammad-ldap-sync

Issues #XXX

### CODE REVIEW

File: lib-spring/src/main/java/de/muenchen/oss/ezldap/spring/rest/v1/LdapUserController.java [hunk 1]
```diff
@@ -179,7 +179,8 @@ public ResponseEntity<List<LdapBaseUserDTO>> searchUser(
                     @ApiResponse(responseCode = "404", description = "OU existiert nicht"),
             }
     )
-    public ResponseEntity<List<LdapBaseUserDTO>> findByOu(@RequestParam(name = "ou") final String ou) {
+    public ResponseEntity<List<LdapBaseUserDTO>> findByOu(@RequestParam(name = "ou") final String ou,
+            @RequestParam(name = "modifyTimeStamp") String modifyTimeStamp) {
```
Comment (FabianWilms) [inline lib-spring/src/main/java/de/muenchen/oss/ezldap/spring/rest/v1/LdapUserController.java:183]: You define RequestParam but then don't use it.

File: lib-core/src/main/java/de/muenchen/oss/ezldap/core/LdapOuNode.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    /**
+     * Creates json representation of the subtree
+     *
+     * @return Json
+     */
+    public String toJson() throws JsonProcessingException {
+        return new ObjectMapper().writeValueAsString(this);
+    }
```
Comment (FabianWilms) [inline lib-core/src/main/java/de/muenchen/oss/ezldap/core/LdapOuNode.java:81]: What do you need this method for? If we remove it we can also remove the dependency on jackson.

File: lib-core/src/main/java/de/muenchen/oss/ezldap/core/LdapOuNode.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+        if (getUsers() != null)
+            getUsers().forEach(u -> tree.append(tab + u.toString() + System.lineSeparator()));
+
+        getChildNodes().forEach((k, v) -> {
+            tree.append(v.toTree(tab + "     "));
+        });
+
+        return tree.toString();
+    }
```
Comment (FabianWilms) [inline lib-core/src/main/java/de/muenchen/oss/ezldap/core/LdapOuNode.java:72]: These methods have no use in this lib and seem more like you need them in your project for debugging purposes? Accordingly, I would actually remove them from this class.

File: lib-core/src/main/java/de/muenchen/oss/ezldap/core/LdapOuNode.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+     */
+    public String toJson() throws JsonProcessingException {
+        return new ObjectMapper().writeValueAsString(this);
+    }
+
+    /**
+     * Creates a map with lhmObjectId and LdapUserDTO for easy finding and access of users contained in
+     * the subtree
+     *
+     * @return map
```
Comment (FabianWilms) [inline lib-core/src/main/java/de/muenchen/oss/ezldap/core/LdapOuNode.java:87]: Add docs how the returned map is structured --> KEY = lhmObjectId, VALUE = LdapUserDto

**Review by FabianWilms — CHANGES_REQUESTED** [review]
Further: You seem to have only added the new features to the Java-Library of ezLdap and not the REST-API. I don't know wether this is something that @eidottermihi is ok with.

File: lib-spring/src/main/java/de/muenchen/oss/ezldap/spring/rest/v1/LdapUserController.java (hunk 1 above)
Comment (sfi2022) [reply lib-spring/src/main/java/de/muenchen/oss/ezldap/spring/rest/v1/LdapUserController.java:183]: That wasn't me. I only removed one parameter that was included in the sources I used.

File: lib-core/src/main/java/de/muenchen/oss/ezldap/core/LdapOuNode.java (hunk 2 above)
Comment (sfi2022) [reply lib-core/src/main/java/de/muenchen/oss/ezldap/core/LdapOuNode.java:81]: As i wrote in the mr comments above it is not necessary. I will remove it.

File: lib-core/src/main/java/de/muenchen/oss/ezldap/core/LdapOuNode.java (hunk 3 above)
Comment (sfi2022) [reply lib-core/src/main/java/de/muenchen/oss/ezldap/core/LdapOuNode.java:72]: If we remove the JSON representation, this is the only possible view to display the created subtree in a human-readable way. For example in log files, if something is going wrong. Even ApacheDirectoryStudio is not showing the the combination of ou and assigned users.

File: lib-core/src/main/java/de/muenchen/oss/ezldap/core/LdapOuNode.java (hunk 4 above)
Comment (sfi2022) [reply lib-core/src/main/java/de/muenchen/oss/ezldap/core/LdapOuNode.java:87]: I remove this. Use flatListLdapUserDTO and build the map by your own.

**Review by FabianWilms — APPROVED** [review]

**Review by FabianWilms — COMMENTED** [review]
Looks good from my end, @eidottermihi do you want to review now?

**Comment by eidottermihi** [comment]
We (@sfi2022 @FabianWilms and @eidottermihi ) discussed this PR and today and we currently don't see the value in adding this functionality to ezLDAP as a "general purpose" library.

The use case from Zammad is rather unique and bringing such an use case upstream does not feel right.

If the future shows that this use case is no longer unique as more projects/services need such a functionality, I'm more than willing to open this topic again.

**Pull request closed without merging**